src/clusterfuzz/_internal/fuzzing/corpus_manager.py

```diff
@@ -603,25 +360,228 @@ def get_regressions_corpus_gcs_url(self):
+    dated_backup_url = gcs_url_for_backup_file(
+        backup_bucket_name, corpus.engine, corpus.project_qualified_target_name,
+        timestamp)
+
+    if not storage.copy_file_to(backup_archive_path, dated_backup_url):
+      return None
+
+    latest_backup_url = gcs_url_for_backup_file(
+        backup_bucket_name, corpus.engine, corpus.project_qualified_target_name,
+        LATEST_BACKUP_TIMESTAMP)
 
-  def __init__(self, *args, **kwargs):
-    super().__init__(*args, **kwargs)
-    self._buf_list = []
+    if not storage.copy_blob(dated_backup_url, latest_backup_url):
+      logs.log_error(
+          'Failed to update latest corpus backup at "%s"' % latest_backup_url)
+  except Exception as ex:
+    logs.log_error(
+        'backup_corpus failed: %s\n' % str(ex),
+        backup_bucket_name=backup_bucket_name,
+        directory=directory,
+        backup_archive_path=backup_archive_path)
 
-  def write(self, data):
-    """Normal behaving write method."""
-    self._buf_list.append(data)
-    return len(data)
+  finally:
+    # Remove backup archive.
+    shell.remove_file(backup_archive_path)
 
-  def destructive_read(self):
-    """Returns all the data we have and discards it."""
-    old_buf = b''.join(self._buf_list)
-    self._buf_list = []
-    return old_buf
+  return dated_backup_url
+
+
+def gcs_url_for_backup_directory(backup_bucket_name, fuzzer_name,
+                                 project_qualified_target_name):
+  """Build GCS URL for corpus backup directory.
 
+  Returns:
+    A string giving the GCS URL.
+  """
+  return (f'gs://{backup_bucket_name}/corpus/{fuzzer_name}/'
+          f'{project_qualified_target_name}/')
+
+
+def _get_regressions_corpus_gcs_url(bucket_name, bucket_path):
+  """Return gcs path to directory containing crash regressions."""
+  return _get_gcs_url(
+      bucket_name, bucket_path, suffix=REGRESSIONS_GCS_PATH_SUFFIX)
+
+
+def download_corpus(corpus, directory):
+  storage.download_signed_urls(corpus.download_urls, directory)
+  storage.download_signed_urls(corpus.regression_download_urls, directory)
+
+
+def _get_gcs_url(bucket_name, bucket_path, suffix=''):
+  """Build corpus GCS URL for gsutil.
+  Returns:
+    A string giving the GCS URL.
+  """
+  # TODO(metzman): Delete this after we are done migrating to the zipcorpus
+  # format.
+  url = f'gs://{bucket_name}{bucket_path}{suffix}'
+  if not url.endswith('/'):
+    # Ensure that the bucket path is '/' terminated. Without this, when a
+    # single file is being uploaded, it is renamed to the trailing non-/
+    # terminated directory name instead.
+    url += '/'
+
+  return url
+
+
+def get_proto_corpus(bucket_name, bucket_path):
+  """Returns a proto representation of a corpus."""
+  gcs_url = _get_gcs_url(bucket_name, bucket_path)
+  corpus_urls = {}
+  # TODO(metzman): Allow this step to be skipped by trusted fuzzers.
+  for corpus_element_url in storage.list_blobs(gcs_url):
```

Comment:
Not sure this would work. It seems to me that `list_blobs` only return names, not full path.

Reply:
You are right.